{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import figen_class"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Requirement already satisfied: lightgbm in /Users/jangsehwan/opt/anaconda3/lib/python3.9/site-packages (2.0.11)\n",
      "Requirement already satisfied: numpy in /Users/jangsehwan/opt/anaconda3/lib/python3.9/site-packages (from lightgbm) (1.21.5)\n",
      "Requirement already satisfied: scipy in /Users/jangsehwan/opt/anaconda3/lib/python3.9/site-packages (from lightgbm) (1.7.3)\n",
      "Requirement already satisfied: scikit-learn in /Users/jangsehwan/opt/anaconda3/lib/python3.9/site-packages (from lightgbm) (1.0.2)\n",
      "Requirement already satisfied: joblib>=0.11 in /Users/jangsehwan/opt/anaconda3/lib/python3.9/site-packages (from scikit-learn->lightgbm) (1.1.0)\n",
      "Requirement already satisfied: threadpoolctl>=2.0.0 in /Users/jangsehwan/opt/anaconda3/lib/python3.9/site-packages (from scikit-learn->lightgbm) (2.2.0)\n",
      "\n",
      "\u001b[1m[\u001b[0m\u001b[34;49mnotice\u001b[0m\u001b[1;39;49m]\u001b[0m\u001b[39;49m A new release of pip is available: \u001b[0m\u001b[31;49m23.1.2\u001b[0m\u001b[39;49m -> \u001b[0m\u001b[32;49m23.2.1\u001b[0m\n",
      "\u001b[1m[\u001b[0m\u001b[34;49mnotice\u001b[0m\u001b[1;39;49m]\u001b[0m\u001b[39;49m To update, run: \u001b[0m\u001b[32;49mpip install --upgrade pip\u001b[0m\n"
     ]
    }
   ],
   "source": [
    "!pip install lightgbm"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Collecting package metadata (current_repodata.json): done\n",
      "Solving environment: done\n",
      "\n",
      "\n",
      "==> WARNING: A newer version of conda exists. <==\n",
      "  current version: 4.13.0\n",
      "  latest version: 23.7.3\n",
      "\n",
      "Please update conda by running\n",
      "\n",
      "    $ conda update -n base -c defaults conda\n",
      "\n",
      "\n",
      "\n",
      "## Package Plan ##\n",
      "\n",
      "  environment location: /Users/jangsehwan/opt/anaconda3\n",
      "\n",
      "  added / updated specs:\n",
      "    - lightgbm\n",
      "\n",
      "\n",
      "The following packages will be downloaded:\n",
      "\n",
      "    package                    |            build\n",
      "    ---------------------------|-----------------\n",
      "    ca-certificates-2023.7.22  |       h8857fd0_0         146 KB  conda-forge\n",
      "    certifi-2023.7.22          |     pyhd8ed1ab_0         150 KB  conda-forge\n",
      "    conda-22.11.1              |   py39h6e9494a_1         874 KB  conda-forge\n",
      "    lightgbm-3.1.1             |   py39h219cf5c_0        1022 KB  conda-forge\n",
      "    openssl-1.1.1v             |       h8a1eda9_0         1.7 MB  conda-forge\n",
      "    python_abi-3.9             |           2_cp39           4 KB  conda-forge\n",
      "    ruamel.yaml-0.17.32        |   py39hdc70f33_0         197 KB  conda-forge\n",
      "    ruamel.yaml.clib-0.2.7     |   py39ha30fb19_1         118 KB  conda-forge\n",
      "    ------------------------------------------------------------\n",
      "                                           Total:         4.1 MB\n",
      "\n",
      "The following NEW packages will be INSTALLED:\n",
      "\n",
      "  lightgbm           conda-forge/osx-64::lightgbm-3.1.1-py39h219cf5c_0\n",
      "  python_abi         conda-forge/osx-64::python_abi-3.9-2_cp39\n",
      "  ruamel.yaml        conda-forge/osx-64::ruamel.yaml-0.17.32-py39hdc70f33_0\n",
      "  ruamel.yaml.clib   conda-forge/osx-64::ruamel.yaml.clib-0.2.7-py39ha30fb19_1\n",
      "\n",
      "The following packages will be UPDATED:\n",
      "\n",
      "  ca-certificates    pkgs/main::ca-certificates-2022.4.26-~ --> conda-forge::ca-certificates-2023.7.22-h8857fd0_0\n",
      "  certifi            pkgs/main/osx-64::certifi-2022.5.18.1~ --> conda-forge/noarch::certifi-2023.7.22-pyhd8ed1ab_0\n",
      "  conda              pkgs/main::conda-4.13.0-py39hecd8cb5_0 --> conda-forge::conda-22.11.1-py39h6e9494a_1\n",
      "  openssl              pkgs/main::openssl-1.1.1o-hca72f7f_0 --> conda-forge::openssl-1.1.1v-h8a1eda9_0\n",
      "\n",
      "\n",
      "Proceed ([y]/n)? ^C\n",
      "\n",
      "CondaSystemExit: \n",
      "Operation aborted.  Exiting.\n",
      "\n"
     ]
    }
   ],
   "source": [
    "!conda install -c conda-forge lightgbm \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 55,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Collecting category_encoders\n",
      "  Downloading category_encoders-2.6.2-py2.py3-none-any.whl (81 kB)\n",
      "\u001b[2K     \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m81.8/81.8 kB\u001b[0m \u001b[31m2.3 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n",
      "\u001b[?25hRequirement already satisfied: numpy>=1.14.0 in /Users/jangsehwan/opt/anaconda3/lib/python3.9/site-packages (from category_encoders) (1.21.5)\n",
      "Requirement already satisfied: scikit-learn>=0.20.0 in /Users/jangsehwan/opt/anaconda3/lib/python3.9/site-packages (from category_encoders) (1.0.2)\n",
      "Requirement already satisfied: scipy>=1.0.0 in /Users/jangsehwan/opt/anaconda3/lib/python3.9/site-packages (from category_encoders) (1.7.3)\n",
      "Requirement already satisfied: statsmodels>=0.9.0 in /Users/jangsehwan/opt/anaconda3/lib/python3.9/site-packages (from category_encoders) (0.13.2)\n",
      "Requirement already satisfied: pandas>=1.0.5 in /Users/jangsehwan/opt/anaconda3/lib/python3.9/site-packages (from category_encoders) (1.4.2)\n",
      "Requirement already satisfied: patsy>=0.5.1 in /Users/jangsehwan/opt/anaconda3/lib/python3.9/site-packages (from category_encoders) (0.5.2)\n",
      "Requirement already satisfied: python-dateutil>=2.8.1 in /Users/jangsehwan/opt/anaconda3/lib/python3.9/site-packages (from pandas>=1.0.5->category_encoders) (2.8.2)\n",
      "Requirement already satisfied: pytz>=2020.1 in /Users/jangsehwan/opt/anaconda3/lib/python3.9/site-packages (from pandas>=1.0.5->category_encoders) (2022.1)\n",
      "Requirement already satisfied: six in /Users/jangsehwan/opt/anaconda3/lib/python3.9/site-packages (from patsy>=0.5.1->category_encoders) (1.16.0)\n",
      "Requirement already satisfied: joblib>=0.11 in /Users/jangsehwan/opt/anaconda3/lib/python3.9/site-packages (from scikit-learn>=0.20.0->category_encoders) (1.1.0)\n",
      "Requirement already satisfied: threadpoolctl>=2.0.0 in /Users/jangsehwan/opt/anaconda3/lib/python3.9/site-packages (from scikit-learn>=0.20.0->category_encoders) (2.2.0)\n",
      "Requirement already satisfied: packaging>=21.3 in /Users/jangsehwan/opt/anaconda3/lib/python3.9/site-packages (from statsmodels>=0.9.0->category_encoders) (21.3)\n",
      "Requirement already satisfied: pyparsing!=3.0.5,>=2.0.2 in /Users/jangsehwan/opt/anaconda3/lib/python3.9/site-packages (from packaging>=21.3->statsmodels>=0.9.0->category_encoders) (3.0.4)\n",
      "Installing collected packages: category_encoders\n",
      "Successfully installed category_encoders-2.6.2\n",
      "\n",
      "\u001b[1m[\u001b[0m\u001b[34;49mnotice\u001b[0m\u001b[1;39;49m]\u001b[0m\u001b[39;49m A new release of pip is available: \u001b[0m\u001b[31;49m23.1.2\u001b[0m\u001b[39;49m -> \u001b[0m\u001b[32;49m23.2.1\u001b[0m\n",
      "\u001b[1m[\u001b[0m\u001b[34;49mnotice\u001b[0m\u001b[1;39;49m]\u001b[0m\u001b[39;49m To update, run: \u001b[0m\u001b[32;49mpip install --upgrade pip\u001b[0m\n"
     ]
    }
   ],
   "source": [
    "!pip install category_encoders"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 56,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\u001b[31mERROR: Could not find a version that satisfies the requirement imblearn.under_sampling (from versions: none)\u001b[0m\u001b[31m\n",
      "\u001b[0m\u001b[31mERROR: No matching distribution found for imblearn.under_sampling\u001b[0m\u001b[31m\n",
      "\u001b[0m\n",
      "\u001b[1m[\u001b[0m\u001b[34;49mnotice\u001b[0m\u001b[1;39;49m]\u001b[0m\u001b[39;49m A new release of pip is available: \u001b[0m\u001b[31;49m23.1.2\u001b[0m\u001b[39;49m -> \u001b[0m\u001b[32;49m23.2.1\u001b[0m\n",
      "\u001b[1m[\u001b[0m\u001b[34;49mnotice\u001b[0m\u001b[1;39;49m]\u001b[0m\u001b[39;49m To update, run: \u001b[0m\u001b[32;49mpip install --upgrade pip\u001b[0m\n",
      "\u001b[31mERROR: Could not find a version that satisfies the requirement imblearn.ensemble (from versions: none)\u001b[0m\u001b[31m\n",
      "\u001b[0m\u001b[31mERROR: No matching distribution found for imblearn.ensemble\u001b[0m\u001b[31m\n",
      "\u001b[0m\n",
      "\u001b[1m[\u001b[0m\u001b[34;49mnotice\u001b[0m\u001b[1;39;49m]\u001b[0m\u001b[39;49m A new release of pip is available: \u001b[0m\u001b[31;49m23.1.2\u001b[0m\u001b[39;49m -> \u001b[0m\u001b[32;49m23.2.1\u001b[0m\n",
      "\u001b[1m[\u001b[0m\u001b[34;49mnotice\u001b[0m\u001b[1;39;49m]\u001b[0m\u001b[39;49m To update, run: \u001b[0m\u001b[32;49mpip install --upgrade pip\u001b[0m\n"
     ]
    }
   ],
   "source": [
    "!pip install imblearn.under_sampling\n",
    "!pip install imblearn.ensemble"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 57,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Unexpected exception formatting exception. Falling back to standard exception\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Traceback (most recent call last):\n",
      "  File \"/Users/jangsehwan/opt/anaconda3/lib/python3.9/site-packages/IPython/core/interactiveshell.py\", line 3397, in run_code\n",
      "    exec(code_obj, self.user_global_ns, self.user_ns)\n",
      "  File \"/var/folders/5k/2xvk5vx160schsvjdh057hj80000gn/T/ipykernel_98377/4159967545.py\", line 4, in <cell line: 4>\n",
      "    from lightgbm import LGBMClassifier\n",
      "  File \"/Users/jangsehwan/opt/anaconda3/lib/python3.9/site-packages/lightgbm/__init__.py\", line 8, in <module>\n",
      "    from .basic import Booster, Dataset\n",
      "  File \"/Users/jangsehwan/opt/anaconda3/lib/python3.9/site-packages/lightgbm/basic.py\", line 15, in <module>\n",
      "    from .compat import (PANDAS_INSTALLED, DataFrame, Series, is_dtype_sparse,\n",
      "ImportError: cannot import name 'PANDAS_INSTALLED' from 'lightgbm.compat' (/Users/jangsehwan/opt/anaconda3/lib/python3.9/site-packages/lightgbm/compat.py)\n",
      "\n",
      "During handling of the above exception, another exception occurred:\n",
      "\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/jangsehwan/opt/anaconda3/lib/python3.9/site-packages/IPython/core/interactiveshell.py\", line 1992, in showtraceback\n",
      "    stb = self.InteractiveTB.structured_traceback(\n",
      "  File \"/Users/jangsehwan/opt/anaconda3/lib/python3.9/site-packages/IPython/core/ultratb.py\", line 1118, in structured_traceback\n",
      "    return FormattedTB.structured_traceback(\n",
      "  File \"/Users/jangsehwan/opt/anaconda3/lib/python3.9/site-packages/IPython/core/ultratb.py\", line 1012, in structured_traceback\n",
      "    return VerboseTB.structured_traceback(\n",
      "  File \"/Users/jangsehwan/opt/anaconda3/lib/python3.9/site-packages/IPython/core/ultratb.py\", line 865, in structured_traceback\n",
      "    formatted_exception = self.format_exception_as_a_whole(etype, evalue, etb, number_of_lines_of_context,\n",
      "  File \"/Users/jangsehwan/opt/anaconda3/lib/python3.9/site-packages/IPython/core/ultratb.py\", line 818, in format_exception_as_a_whole\n",
      "    frames.append(self.format_record(r))\n",
      "  File \"/Users/jangsehwan/opt/anaconda3/lib/python3.9/site-packages/IPython/core/ultratb.py\", line 736, in format_record\n",
      "    result += ''.join(_format_traceback_lines(frame_info.lines, Colors, self.has_colors, lvals))\n",
      "  File \"/Users/jangsehwan/opt/anaconda3/lib/python3.9/site-packages/stack_data/utils.py\", line 145, in cached_property_wrapper\n",
      "    value = obj.__dict__[self.func.__name__] = self.func(obj)\n",
      "  File \"/Users/jangsehwan/opt/anaconda3/lib/python3.9/site-packages/stack_data/core.py\", line 698, in lines\n",
      "    pieces = self.included_pieces\n",
      "  File \"/Users/jangsehwan/opt/anaconda3/lib/python3.9/site-packages/stack_data/utils.py\", line 145, in cached_property_wrapper\n",
      "    value = obj.__dict__[self.func.__name__] = self.func(obj)\n",
      "  File \"/Users/jangsehwan/opt/anaconda3/lib/python3.9/site-packages/stack_data/core.py\", line 649, in included_pieces\n",
      "    pos = scope_pieces.index(self.executing_piece)\n",
      "  File \"/Users/jangsehwan/opt/anaconda3/lib/python3.9/site-packages/stack_data/utils.py\", line 145, in cached_property_wrapper\n",
      "    value = obj.__dict__[self.func.__name__] = self.func(obj)\n",
      "  File \"/Users/jangsehwan/opt/anaconda3/lib/python3.9/site-packages/stack_data/core.py\", line 628, in executing_piece\n",
      "    return only(\n",
      "  File \"/Users/jangsehwan/opt/anaconda3/lib/python3.9/site-packages/executing/executing.py\", line 164, in only\n",
      "    raise NotOneValueFound('Expected one value, found 0')\n",
      "executing.executing.NotOneValueFound: Expected one value, found 0\n"
     ]
    }
   ],
   "source": [
    "\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "from sklearn.model_selection import train_test_split\n",
    "from lightgbm import LGBMClassifier\n",
    "from sklearn.metrics import classification_report, roc_auc_score, f1_score\n",
    "from sklearn.preprocessing import OneHotEncoder\n",
    "from category_encoders import TargetEncoder, WOEEncoder\n",
    "from imblearn.under_sampling import RandomUnderSampler\n",
    "from imblearn.ensemble import BalancedRandomForestClassifier\n",
    "from sdv.single_table import CTGANSynthesizer\n",
    "import figen_class \n",
    "\n",
    "\n",
    "\n",
    "class imblanced_compare:\n",
    "    \n",
    "    def __init__(self, data_path, categorical_columns:list):\n",
    "        self.data_path = data_path\n",
    "        self.categorical_columns = categorical_columns\n",
    "\n",
    "    def load_dataset(file_path):\n",
    "        data = pd.read_csv(file_path)\n",
    "        X = data.drop(columns=['target'])\n",
    "        y = data['target']\n",
    "        return X, y\n",
    "\n",
    "    def split_data(X, y, test_size=0.2, random_state=1004):\n",
    "        return train_test_split(X, y, test_size=test_size, random_state=random_state)\n",
    "    \n",
    "    def under_sampling(X_train, y_train):\n",
    "        rus = RandomUnderSampler()\n",
    "        X_resampled, y_resampled = rus.fit_resample(X_train, y_train)\n",
    "        return X_resampled, y_resampled\n",
    "    \n",
    "    def ensemble_under_sampling(X_train, y_train):\n",
    "        brf = BalancedRandomForestClassifier()\n",
    "        brf.fit(X_train, y_train)\n",
    "        return brf\n",
    "    \n",
    "    def generate_synthetic_data(self,X_resampled, y_resampled):\n",
    "        figen = FiGen(ratio=0.3, categorical_columns=self.categorical_columns)\n",
    "        synthetic_X, synthetic_y = figen.fit(X_resampled, y_resampled, X_train, y_train)\n",
    "        return synthetic_X, synthetic_y\n",
    "    \n",
    "    def encode_data(X_train, y_train):\n",
    "        ohe = OneHotEncoder()\n",
    "        X_ohe = ohe.fit_transform(X_train[categorical_columns])\n",
    "    \n",
    "        mean_encoder = TargetEncoder()\n",
    "        X_mean_encoded = mean_encoder.fit_transform(X_train[categorical_columns], y_train)\n",
    "    \n",
    "        woe_encoder = WOEEncoder()\n",
    "        X_woe_encoded = woe_encoder.fit_transform(X_train[categorical_columns], y_train)\n",
    "    \n",
    "        return X_ohe, X_mean_encoded, X_woe_encoded\n",
    "    \n",
    "    def evaluate_model(model, X_train, y_train, X_test, y_test):\n",
    "        model.fit(X_train, y_train)\n",
    "        y_pred = model.predict(X_test)\n",
    "        roc_auc = roc_auc_score(y_test, y_pred)\n",
    "        f1 = f1_score(y_test, y_pred)\n",
    "        return roc_auc, f1, classification_report(y_test, y_pred)\n",
    "    \n",
    "    def fit(self,data, categorical_columns:list):\n",
    "        # Load and preprocess data\n",
    "        X, y = self.load_dataset(data)\n",
    "        X_train, X_test, y_train, y_test = split_data(X, y)\n",
    "    \n",
    "        # Create LGBM model\n",
    "        lgbm_model = LGBMClassifier()\n",
    "    \n",
    "        # Under Sampling\n",
    "        X_resampled, y_resampled = self.under_sampling(X_train, y_train)\n",
    "        roc_auc_us, f1_us, report_us = self.evaluate_model(lgbm_model, X_resampled, y_resampled, X_test, y_test)\n",
    "    \n",
    "        # Ensemble with Under Sampling\n",
    "        brf = self.ensemble_under_sampling(X_train, y_train)\n",
    "        roc_auc_brf, f1_brf, report_brf = self.evaluate_model(brf, X_train, y_train, X_test, y_test)\n",
    "    \n",
    "        # Generate Synthetic Data\n",
    "        synthetic_X, synthetic_y = self.generate_synthetic_data(X_resampled, y_resampled)\n",
    "        roc_auc_syn, f1_syn, report_syn = self.evaluate_model(lgbm_model, synthetic_X, synthetic_y, X_test, y_test)\n",
    "    \n",
    "        # Encode Data\n",
    "        X_ohe, X_mean_encoded, X_woe_encoded = self.encode_data(X_train, y_train)\n",
    "        roc_auc_ohe, f1_ohe, report_ohe = self.evaluate_model(lgbm_model, X_ohe, y_train, X_test, y_test)\n",
    "        roc_auc_mean_encoded, f1_mean_encoded, report_mean_encoded = self.evaluate_model(lgbm_model, X_mean_encoded, y_train, X_test, y_test)\n",
    "        roc_auc_woe_encoded, f1_woe_encoded, report_woe_encoded = self.evaluate_model(lgbm_model, X_woe_encoded, y_train, X_test, y_test)\n",
    "    \n",
    "        # Print Results\n",
    "        print(\"Under Sampling Results:\")\n",
    "        print(\"ROC AUC:\", roc_auc_us)\n",
    "        print(\"F1 Score:\", f1_us)\n",
    "        print(report_us)\n",
    "    \n",
    "        print(\"\\nBalanced Random Forest Results:\")\n",
    "        print(\"ROC AUC:\", roc_auc_brf)\n",
    "        print(\"F1 Score:\", f1_brf)\n",
    "        print(report_brf)\n",
    "    \n",
    "        print(\"\\nSynthetic Data (FiGen) Results:\")\n",
    "        print(\"ROC AUC:\", roc_auc_syn)\n",
    "        print(\"F1 Score:\", f1_syn)\n",
    "        print(report_syn)\n",
    "    \n",
    "        print(\"\\nOne-Hot Encoding Results:\")\n",
    "        print(\"ROC AUC:\", roc_auc_ohe)\n",
    "        print(\"F1 Score:\", f1_ohe)\n",
    "        print(report_ohe)\n",
    "    \n",
    "        print(\"\\nMean Encoding Results:\")\n",
    "        print(\"ROC AUC:\", roc_auc_mean_encoded)\n",
    "        print(\"F1 Score:\", f1_mean_encoded)\n",
    "        print(report_mean_encoded)\n",
    "    \n",
    "        print(\"\\nWOE Encoding Results:\")\n",
    "        print(\"ROC AUC:\", roc_auc_woe_encoded)\n",
    "        print(\"F1 Score:\", f1_woe_encoded)\n",
    "        print(report_woe_encoded)\n",
    "    \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 58,
   "metadata": {},
   "outputs": [
    {
     "ename": "NameError",
     "evalue": "name 'ImbalancedCompare' is not defined",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mNameError\u001b[0m                                 Traceback (most recent call last)",
      "\u001b[1;32m/Users/jangsehwan/Financial-Imbalanced-Data-With-Deep-Density-Hybrid-Sampling/imbalanced-ddhs/scripts/model_test_v0.00.01.ipynb 셀 7\u001b[0m in \u001b[0;36m<cell line: 1>\u001b[0;34m()\u001b[0m\n\u001b[0;32m----> <a href='vscode-notebook-cell:/Users/jangsehwan/Financial-Imbalanced-Data-With-Deep-Density-Hybrid-Sampling/imbalanced-ddhs/scripts/model_test_v0.00.01.ipynb#W3sZmlsZQ%3D%3D?line=0'>1</a>\u001b[0m test \u001b[39m=\u001b[39m ImbalancedCompare(data_path\u001b[39m=\u001b[39m\u001b[39m'\u001b[39m\u001b[39myour_data.csv\u001b[39m\u001b[39m'\u001b[39m, categorical_columns\u001b[39m=\u001b[39m[\u001b[39m'\u001b[39m\u001b[39mcat_col1\u001b[39m\u001b[39m'\u001b[39m, \u001b[39m'\u001b[39m\u001b[39mcat_col2\u001b[39m\u001b[39m'\u001b[39m])\n\u001b[1;32m      <a href='vscode-notebook-cell:/Users/jangsehwan/Financial-Imbalanced-Data-With-Deep-Density-Hybrid-Sampling/imbalanced-ddhs/scripts/model_test_v0.00.01.ipynb#W3sZmlsZQ%3D%3D?line=1'>2</a>\u001b[0m test\u001b[39m.\u001b[39mfit()\n",
      "\u001b[0;31mNameError\u001b[0m: name 'ImbalancedCompare' is not defined"
     ]
    }
   ],
   "source": [
    "test = ImbalancedCompare(data_path='your_data.csv', categorical_columns=['cat_col1', 'cat_col2'])\n",
    "test.fit()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 54,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Unexpected exception formatting exception. Falling back to standard exception\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Traceback (most recent call last):\n",
      "  File \"/Users/jangsehwan/opt/anaconda3/lib/python3.9/site-packages/IPython/core/interactiveshell.py\", line 3397, in run_code\n",
      "    exec(code_obj, self.user_global_ns, self.user_ns)\n",
      "  File \"/var/folders/5k/2xvk5vx160schsvjdh057hj80000gn/T/ipykernel_98377/3204302439.py\", line 5, in <cell line: 5>\n",
      "    from lightgbm import LGBMClassifier\n",
      "  File \"/Users/jangsehwan/opt/anaconda3/lib/python3.9/site-packages/lightgbm/__init__.py\", line 8, in <module>\n",
      "    from .basic import Booster, Dataset\n",
      "  File \"/Users/jangsehwan/opt/anaconda3/lib/python3.9/site-packages/lightgbm/basic.py\", line 15, in <module>\n",
      "    from .compat import (PANDAS_INSTALLED, DataFrame, Series, is_dtype_sparse,\n",
      "ImportError: cannot import name 'PANDAS_INSTALLED' from 'lightgbm.compat' (/Users/jangsehwan/opt/anaconda3/lib/python3.9/site-packages/lightgbm/compat.py)\n",
      "\n",
      "During handling of the above exception, another exception occurred:\n",
      "\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/jangsehwan/opt/anaconda3/lib/python3.9/site-packages/IPython/core/interactiveshell.py\", line 1992, in showtraceback\n",
      "    stb = self.InteractiveTB.structured_traceback(\n",
      "  File \"/Users/jangsehwan/opt/anaconda3/lib/python3.9/site-packages/IPython/core/ultratb.py\", line 1118, in structured_traceback\n",
      "    return FormattedTB.structured_traceback(\n",
      "  File \"/Users/jangsehwan/opt/anaconda3/lib/python3.9/site-packages/IPython/core/ultratb.py\", line 1012, in structured_traceback\n",
      "    return VerboseTB.structured_traceback(\n",
      "  File \"/Users/jangsehwan/opt/anaconda3/lib/python3.9/site-packages/IPython/core/ultratb.py\", line 865, in structured_traceback\n",
      "    formatted_exception = self.format_exception_as_a_whole(etype, evalue, etb, number_of_lines_of_context,\n",
      "  File \"/Users/jangsehwan/opt/anaconda3/lib/python3.9/site-packages/IPython/core/ultratb.py\", line 818, in format_exception_as_a_whole\n",
      "    frames.append(self.format_record(r))\n",
      "  File \"/Users/jangsehwan/opt/anaconda3/lib/python3.9/site-packages/IPython/core/ultratb.py\", line 736, in format_record\n",
      "    result += ''.join(_format_traceback_lines(frame_info.lines, Colors, self.has_colors, lvals))\n",
      "  File \"/Users/jangsehwan/opt/anaconda3/lib/python3.9/site-packages/stack_data/utils.py\", line 145, in cached_property_wrapper\n",
      "    value = obj.__dict__[self.func.__name__] = self.func(obj)\n",
      "  File \"/Users/jangsehwan/opt/anaconda3/lib/python3.9/site-packages/stack_data/core.py\", line 698, in lines\n",
      "    pieces = self.included_pieces\n",
      "  File \"/Users/jangsehwan/opt/anaconda3/lib/python3.9/site-packages/stack_data/utils.py\", line 145, in cached_property_wrapper\n",
      "    value = obj.__dict__[self.func.__name__] = self.func(obj)\n",
      "  File \"/Users/jangsehwan/opt/anaconda3/lib/python3.9/site-packages/stack_data/core.py\", line 649, in included_pieces\n",
      "    pos = scope_pieces.index(self.executing_piece)\n",
      "  File \"/Users/jangsehwan/opt/anaconda3/lib/python3.9/site-packages/stack_data/utils.py\", line 145, in cached_property_wrapper\n",
      "    value = obj.__dict__[self.func.__name__] = self.func(obj)\n",
      "  File \"/Users/jangsehwan/opt/anaconda3/lib/python3.9/site-packages/stack_data/core.py\", line 628, in executing_piece\n",
      "    return only(\n",
      "  File \"/Users/jangsehwan/opt/anaconda3/lib/python3.9/site-packages/executing/executing.py\", line 164, in only\n",
      "    raise NotOneValueFound('Expected one value, found 0')\n",
      "executing.executing.NotOneValueFound: Expected one value, found 0\n"
     ]
    }
   ],
   "source": [
    "import numpy as np\n",
    "import pandas as pd\n",
    "from typing import List, Tuple\n",
    "from sklearn.model_selection import train_test_split\n",
    "from lightgbm import LGBMClassifier\n",
    "from sklearn.metrics import classification_report, roc_auc_score, f1_score\n",
    "from sklearn.preprocessing import OneHotEncoder\n",
    "from category_encoders import TargetEncoder, WOEEncoder\n",
    "from imblearn.under_sampling import RandomUnderSampler\n",
    "from imblearn.ensemble import BalancedRandomForestClassifier\n",
    "from sdv.single_table import CTGANSynthesizer\n",
    "from figen_class import FiGen  # Assuming FiGen is imported from figen_class\n",
    "\n",
    "class ImbalancedCompare:\n",
    "    \n",
    "    def __init__(self, data_path: str, categorical_columns: List[str], generation_rate: float):\n",
    "        self.data_path = data_path\n",
    "        self.categorical_columns = categorical_columns\n",
    "        self.generation_rate = generation_rate\n",
    "\n",
    "    def load_dataset(self, file_path: str) -> Tuple[pd.DataFrame, pd.Series]:\n",
    "        \"\"\"\n",
    "        Load and preprocess the dataset from the given file path.\n",
    "\n",
    "        Args:\n",
    "            file_path (str): Path to the CSV dataset file.\n",
    "\n",
    "        Returns:\n",
    "            X (pd.DataFrame): Feature matrix.\n",
    "            y (pd.Series): Target variable.\n",
    "        \"\"\"\n",
    "        data = pd.read_csv(file_path)\n",
    "        X = data.drop(columns=['target'])\n",
    "        y = data['target']\n",
    "        return X, y\n",
    "\n",
    "    def split_data(self, X: pd.DataFrame, y: pd.Series, test_size: float = self.generation_rate, random_state: int = 1004) -> Tuple[pd.DataFrame, pd.DataFrame, pd.Series, pd.Series]:\n",
    "        \"\"\"\n",
    "        Split the dataset into train and test sets.\n",
    "\n",
    "        Args:\n",
    "            X (pd.DataFrame): Feature matrix.\n",
    "            y (pd.Series): Target variable.\n",
    "            test_size (float): Proportion of the dataset to include in the test split.\n",
    "            random_state (int): Seed for random number generator.\n",
    "\n",
    "        Returns:\n",
    "            X_train (pd.DataFrame): Training features.\n",
    "            X_test (pd.DataFrame): Testing features.\n",
    "            y_train (pd.Series): Training target.\n",
    "            y_test (pd.Series): Testing target.\n",
    "        \"\"\"\n",
    "        return train_test_split(X, y, test_size=test_size, random_state=random_state)\n",
    "    \n",
    "    def under_sampling(self, X_train: pd.DataFrame, y_train: pd.Series) -> Tuple[pd.DataFrame, pd.Series]:\n",
    "        \"\"\"\n",
    "        Apply random under-sampling to balance the classes.\n",
    "\n",
    "        Args:\n",
    "            X_train (pd.DataFrame): Training features.\n",
    "            y_train (pd.Series): Training target.\n",
    "\n",
    "        Returns:\n",
    "            X_resampled (pd.DataFrame): Resampled features.\n",
    "            y_resampled (pd.Series): Resampled target.\n",
    "        \"\"\"\n",
    "        rus = RandomUnderSampler()\n",
    "        X_resampled, y_resampled = rus.fit_resample(X_train, y_train)\n",
    "        return X_resampled, y_resampled\n",
    "    \n",
    "    # ... (Other functions)\n",
    "\n",
    "    def generate_synthetic_data(self, X_resampled: pd.DataFrame, y_resampled: pd.Series) -> Tuple[pd.DataFrame, pd.Series]:\n",
    "        \"\"\"\n",
    "        Generate synthetic data using FiGen.\n",
    "\n",
    "        Args:\n",
    "            X_resampled (pd.DataFrame): Resampled features.\n",
    "            y_resampled (pd.Series): Resampled target.\n",
    "\n",
    "        Returns:\n",
    "            synthetic_X (pd.DataFrame): Synthetic features.\n",
    "            synthetic_y (pd.Series): Synthetic target.\n",
    "        \"\"\"\n",
    "        figen = FiGen(ratio=0.3, categorical_columns=self.categorical_columns)\n",
    "        synthetic_X, synthetic_y = figen.fit(X_resampled, y_resampled, X_train, y_train)\n",
    "        return synthetic_X, synthetic_y\n",
    "    \n",
    "    \n",
    "    def fit(self, data: str):\n",
    "        \"\"\"\n",
    "        Fit the model and evaluate various techniques.\n",
    "\n",
    "        Args:\n",
    "            data (str): Path to the dataset file.\n",
    "        \"\"\"\n",
    "        # Load and preprocess data\n",
    "        X, y = self.load_dataset(data)\n",
    "        X_train, X_test, y_train, y_test = self.split_data(X, y)\n",
    "    \n",
    "        # Create LGBM model\n",
    "        lgbm_model = LGBMClassifier()\n",
    "    \n",
    "        # Under Sampling\n",
    "        X_resampled, y_resampled = self.under_sampling(X_train, y_train)\n",
    "        roc_auc_us, f1_us, report_us = self.evaluate_model(lgbm_model, X_resampled, y_resampled, X_test, y_test)\n",
    "    \n",
    "        # Ensemble with Under Sampling\n",
    "        brf = self.ensemble_under_sampling(X_train, y_train)\n",
    "        roc_auc_brf, f1_brf, report_brf = self.evaluate_model(brf, X_train, y_train, X_test, y_test)\n",
    "    \n",
    "        # Generate Synthetic Data\n",
    "        synthetic_X, synthetic_y = self.generate_synthetic_data(X_resampled, y_resampled)\n",
    "        roc_auc_syn, f1_syn, report_syn = self.evaluate_model(lgbm_model, synthetic_X, synthetic_y, X_test, y_test)\n",
    "    \n",
    "         # Encode Data\n",
    "        X_ohe, X_mean_encoded, X_woe_encoded = self.encode_data(X_train, y_train)\n",
    "        roc_auc_ohe, f1_ohe, report_ohe = self.evaluate_model(lgbm_model, X_ohe, y_train, X_test, y_test)\n",
    "        roc_auc_mean_encoded, f1_mean_encoded, report_mean_encoded = self.evaluate_model(lgbm_model, X_mean_encoded, y_train, X_test, y_test)\n",
    "        roc_auc_woe_encoded, f1_woe_encoded, report_woe_encoded = self.evaluate_model(lgbm_model, X_woe_encoded, y_train, X_test, y_test)\n",
    "    \n",
    "        # Print Results\n",
    "        print(\"Under Sampling Results:\")\n",
    "        print(\"ROC AUC:\", roc_auc_us)\n",
    "        print(\"F1 Score:\", f1_us)\n",
    "        print(report_us)\n",
    "    \n",
    "        print(\"\\nBalanced Random Forest Results:\")\n",
    "        print(\"ROC AUC:\", roc_auc_brf)\n",
    "        print(\"F1 Score:\", f1_brf)\n",
    "        print(report_brf)\n",
    "    \n",
    "        print(\"\\nSynthetic Data (FiGen) Results:\")\n",
    "        print(\"ROC AUC:\", roc_auc_syn)\n",
    "        print(\"F1 Score:\", f1_syn)\n",
    "        print(report_syn)\n",
    "    \n",
    "        print(\"\\nOne-Hot Encoding Results:\")\n",
    "        print(\"ROC AUC:\", roc_auc_ohe)\n",
    "        print(\"F1 Score:\", f1_ohe)\n",
    "        print(report_ohe)\n",
    "    \n",
    "        print(\"\\nMean Encoding Results:\")\n",
    "        print(\"ROC AUC:\", roc_auc_mean_encoded)\n",
    "        print(\"F1 Score:\", f1_mean_encoded)\n",
    "        print(report_mean_encoded)\n",
    "    \n",
    "        print(\"\\nWOE Encoding Results:\")\n",
    "        print(\"ROC AUC:\", roc_auc_woe_encoded)\n",
    "        print(\"F1 Score:\", f1_woe_encoded)\n",
    "        print(report_woe_encoded)\n",
    "    \n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "base",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.7"
  },
  "orig_nbformat": 4
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
